---
title: "RNA161024PS RNAseq Analysis"
author: "Joey Estabrook & Julja Burchard"
date: "February 17, 2017"
output: html_document
---

analysis.R runs prebuilt functions on RNAseq alignments

R session should be run in /proj/<projectname>/results/STAR/ directory or setwd("/proj/<projectname>/results/STAR/") while in a current R session.

### Import libraries and CompBio Repos:
- AbundanceFunctions
- AssociationFunctions
- BcorePlotting
- GenomicsFunctions

This should be a separate updated but static copy of the repo so you are running and testing your script on a static instance. That way drastic updates will not effect your current analysis.

```{r}
require(data.table)
require(NMF)
require(affy)
require(limma)
require(AnnotationDbi)
source("../../code/AbundanceFunctions/BiasReduce.R")  
source("../../code/AbundanceFunctions/ExtractTransformLoad.R")
source("../../code/GenomicsFunctions/ReadAndParse.R")
source("../../code/AssociationFunctions/gs.wrapper.R")  #New gs.wrapper with ags. function
source("../../code/BcorePlotting/SummaryPlots.R")
source("../../code/BcorePlotting/MultipleTestingCorrection.R")
source("../../code/BcorePlotting/ClusteringPlots.R")
source("../../code/AbundanceFunctions/DifferentialAnalysis.R")
source("../../code/AbundanceFunctions/NonVisualOutput.R")
source("../../code/GenomicsFunctions/ReadAndParse.R")
```

### Constants
Set some constants for later use. 

```{r constants}
# constants
hclust.limit = 2^16 # max data rows for hclust
hc.frac.cut = 0.75; # quantile of data distribution reqd in one group's worth of data, if too many rows for hclust() show minimal expression
SJ.counts.na.frac = 0.25;     # max fraction of samples not having detected a splice junction for the splice junction to be retained in raw data
# regression parameters
na.lim = 0 # max NAs per row tolerated by lm() at least in some cases
do.not.regress = "alograw" # control norm not to be used for regression stats
# plotting colors
colors.rgb = c(rgb(0,0,0),rgb(0.1,0.1,1),rgb(0,.7,.7),rgb(0,.7,0),rgb(.7,1,0),rgb(.7,0,.7)) 
```

###Project Annotation
Metadata if provided are found in `md.file`
A basic metadata parser is provided below. Sample IDs are expected either as column headers in a by-column sample orientation, or in a designated column `md.IDcol` in a by-row sample orientation.
## Gene file annotation.
Make sure you have the proper GTF used for alignment and the correct release.
TaxID can be found www.ncbi.nlm.nih.gov/Taxonomy/TaxIdentifier/tax_identifier.cgi
## Optional prior data
If lists of genes of interest with or without additional data or annotation are available, please read them in below.

```{r}
# optional sample associated information (metadata) 
md.file = "/home/exacloud/lustre1/BioCoders/ProjectCollaborations/PSchedin_RNA161024PS/data/sample.assoc.info/RNA_analysis_for_ERpositive.negative_cases_2016.10.txt"
md.orientation = "byRow" # sampleIDs are in @ row. alt:byCol (IDs in @ col)
md.IDcol = "SampleID" # reqd if md.orientation is byRow; byCol==headers are IDs

# optional gene lists
priordata.dir = "/home/exacloud/lustre1/BioCoders/ProjectCollaborations/PSchedin_RNA161024PS/data/genelists/" # directory with prior data files
priordata.pattern = "*.txt"   # pattern to recognize files in dir()
priordata.geneID = "ENSG"     # column name for gene IDs matching STAR's

# gene annotation
taxID=9606
gene2ENSfile="/home/exacloud/lustre1/BioCoders/DataResources/AnnotationSources/ncbi/gene2ensembl.gz"
gene2ENS.col = c("taxID","EntrezID","Gene","RefSeqTranscript","EnsemblTranscript","RefSeqProtein","EnsemblProtein")
gtfFile = "/home/exacloud/lustre1/BioCoders/DataResources/Genomes/hg38/release-87/gtf/Homo_sapiens.GRCh38.87.gtf"
gtf.feature = "gene"
gtf.orig.col = c("gene_id","gene_name","gene_biotype")
gtf.col = c("Gene","Symbol","biotype")
```


### Setup for reading in STAR output

Project directory where star output was stored is specified by `readdir`.
Extract the names of the samples from the output names.
Use `readpattern` to specifiy the star output to read in usually all the files, depends on groups.
Use `label.from.colname` to extract *unique* identifiers of each sample from dirnames.


```{r}
# set up to read in STAR alignments
readdir = "/home/exacloud/lustre1/BioCoders/ProjectCollaborations/PSchedin_RNA161024PS/results/STAR_gene/"
readpattern='^RNA1'
useme.cols = 'RNA1'
label.from.colname='^.*?RNA[^_]*?_([^_]+)_.*?R1_001' #this is unique
samps = dir(path=readdir, pattern=readpattern)
# trimmed sample labels
samp.labels = gsub(label.from.colname,'\\1', samps)  

```

This helps annotate your STAR.data object. These are the "conditions"/"labels"/"groups" for your sample names, if FASTQ filenames contain this information. Use 'annCol.label' as a regex to extract condition identifiers from filenames. There can be multiple conditions if multiple matches are recovered () in the regex. The names of each condition in order recovered from the regex are specified in "names". Set "lm.By" to the names of the condition(s) you intend to use for differential expresion.
If FASTQ filenames don't contain condition(s), bring them in from the metadata file in a later step and use a placeholder for "labels" now, like sample names.

```{r}
annCol.names='group'
annCol.lmBy="ER Status"
# FASTQ filenames are missing treatment groups.  Remake annCol from md.file.
annCol.label = '^.*?RNA[^_]*?_([^_]+)_.*?R1_001' # samp labels for now
```

### Read in STAR output *finally*
Creates STAR.data "list" and writes to current directory (results/STAR) the following csv files:
- raw.gene.counts.FwdReads.csv
- raw.gene.counts.Reads.csv
- raw.gene.counts.RevReads.csv
- raw.SJ.counts.UniqueReads.csv
The gene counts files cover various strandedness states of RNAseq libraries.

```{r}
# read in STAR alignments
STAR.data = read_STAR(useme.cols=useme.cols,label.from.colname=label.from.colname,annCol.label=annCol.label,annCol.names=annCol.names,annCol.normBy=NULL,annCol.lmBy=annCol.lmBy,readpattern=readpattern,unstranded.col = list(gene.counts = c(1:4),SJ.counts = c(1:3,7)))

### move this into read_STAR! ###
# filter out SJs with too many NAs
if(any( names(STAR.data$LoM.raw)=="SJ.counts" & exists("SJ.counts.na.frac") )){
  STAR.data$SJ.counts.orig = STAR.data$LoM.raw$SJ.counts
  for(tag in names(STAR.data$LoM.raw$SJ.counts) ){
    STAR.data$LoM.raw$SJ.counts[[tag]] = 
      STAR.data$LoM.raw$SJ.counts[[tag]][ 
        rowSums(is.na(STAR.data$LoM.raw$SJ.counts[[tag]]))
      <= (SJ.counts.na.frac*ncol(STAR.data$LoM.raw$SJ.counts[[tag]])), ]
  }
}
```

* STAR.data structure=> list of list of list of list of matrices*

  <blockquote> STAR.data[
    expt.design,
    myreads,
    LoM.raw[
      gene.counts[
        Reads(matrix),
        FwdReads(matrix),
        RevReads(matrix)
      ],
      SJ.counts[
      UniqueReads(matrix)
      ]
    ]
  ]</blockquote>

The list `expt.design` is the variable that holds the labels/groups/conditions of your project as extracted from FASTQ filenames. If the FASTQ filenames don't have all the appropriate information, please replace this object with information from your metadata file, `md.file`.

```{r}
#annCol=STAR.data[["expt.design"]]
# will bring in from md.file instead in this case

```


###Parse GTF used for genomic feature identification

```{r}
gtf = readENSgtf(filename=gtfFile)
genes.gtf = gtf[feature==gtf.feature, mget(gtf.orig.col)]
names(genes.gtf) = gtf.col
setkeyv(genes.gtf,gtf.col[1])
##### parse of NCBI's EntrezID to Ensembl translation
Entrez2Ensembl = fread(paste("zgrep",paste0("-E '^",taxID,"'"),gene2ENSfile))
names(Entrez2Ensembl) = gene2ENS.col
setkeyv(Entrez2Ensembl,gene2ENS.col[3])
##### add EntrezIDs to genes.gtf
tmp = Entrez2Ensembl[,mget(gene2ENS.col[2:3])]; tmp=tmp[!duplicated(tmp),]
genes.gtf = merge(genes.gtf,tmp,all.x=TRUE)
rm(tmp)
```

### Bias Reduction
uses AbundanceFunctions/processData.R
output normalized csv files in results/STAR directory
- lowess.gene.counts.csv
- lowess.SJ.counts.csv
- qspln.gene.counts.csv
- qspln.SJ.counts.csv
- quant.gene.counts.csv
- quant.SJ.counts.csv

```{r}
myreads = STAR.data$myreads
LoM.norms = vector(mode='list',length=length(STAR.data$LoM.raw))
names(LoM.norms) = names(STAR.data$LoM.raw)
for(tag in names(STAR.data$LoM.raw) ){
  if( length(STAR.data$LoM.raw[[tag]])>1 ){
    LoM.norms[[tag]] = normMatrix(tag=tag,raw.mat=STAR.data$LoM.raw[[tag]][[myreads]])
  } else {
    LoM.norms[[tag]] = normMatrix(tag=tag,raw.mat=STAR.data$LoM.raw[[tag]][[1]])
  }
}
```


### Optionally read in metadata file
This will create a 2nd metadata struture in addition to STAR.data$expt.design


```{r}
## custom: reannotate from metadata
## run if necessary metadata is in file and not also in FASTQ file names
md.dt = fread(md.file)
# map annotation to read matrix
if( md.orientation == "byCol" ){ # samples are one per column
  idx2 = match( colnames(STAR.data$LoM.raw[[1]][[myreads]]), names(md.dt) )
} else { # samples are one per row
  idx2 = match( colnames(STAR.data$LoM.raw[[1]][[myreads]]), md.dt[,get(md.IDcol)] )
}
idx1 = which(!is.na(idx2)); idx2 = idx2[idx1]
if( sum(idx1==1:ncol(STAR.data$LoM.raw[[1]][[myreads]]))==ncol(STAR.data$LoM.raw[[1]][[myreads]]) ){
  if( md.orientation == "byCol" ){ # samples are one per column
    annCol = NULL
    namecol = setdiff( 1:ncol(md.dt), idx2 )
    md.factors = as.character(md.dt[,namecol,with=F])
    for(k in 1:nrow(md.dt) ){
      annCol[[ md.dt[k,get(names(md.dt)[namecol])] ]][idx1] = 
      as.vector( md.dt[ k, mget(names(md.dt)[idx2]) ] )
    }
  } else { # samples are one per row
    annCol = NULL
    namecol = setdiff( names(md.dt), md.IDcol )
    md.factors = as.character(namecol)
    for( k in setdiff(names(md.dt),md.IDcol) ){
      annCol[[ k ]][idx1] = 
      as.vector( md.dt[ idx2, get(k) ] )
    }
  }
} else {
  stop(paste("Some samples have no annotation in",md.file))
}
```


### Optionally read in gene list files
This will create a list of data tables


```{r}

priordata_v = dir(path=priordata.dir, pattern=priordata.pattern) 
priordata_lsdt = NULL
for( fil in priordata_v ){
  myfil = file.path(priordata.dir,fil,fsep=if(grepl(paste0(.Platform$file.sep,"$"),priordata.dir)){''}else{.Platform$file.sep})
  priordata_lsdt[[sub('.txt$','',fil)]] = fread(myfil)
}

```


###Produce QC plots
#summary.plots 
#currently wraps a boxplot (1), scatterplot (2), density plot (3), sd vs. intensity plot (4), and MA plot (5). The MA plot #behavior plots all sample columns in a matrix vs. one BaseSample (Defaults to False).

#qc.clusters
# Plots for each design factor (default) or for factors specified in facSel
#   1) MDS plot of samples based on all data rows
#   2) heatmaps of the most variable rows (row SDs > overall SD)

```{rp
mytypes = names(LoM.norms)
mynorms = names(LoM.norms[[1]])
# samp.labels is already set above -- no need to redo if set correctly
#samp.labels = gsub('^[^.]+\\.','',colnames(LoM.norms[[1]][[1]]))
grpBy = annCol [[annCol.lmBy]]
annCol.plotme = c(annCol.lmBy,"DV200","RNA Yield (ng)")
clim.pct=0.96
histbins=20
for(i in 1:length(mytypes) ){
  for(j in 1:length(mynorms)){
    # set up matrices and config for plotting
    if( length(STAR.data$LoM.raw[[ mytypes[i] ]])>1 ){
      rawmat = STAR.data$LoM.raw[[ mytypes[i] ]][[myreads]]
    } else {
      rawmat = STAR.data$LoM.raw[[ mytypes[i] ]][[1]]
    }
    normmat= LoM.norms[[ mytypes[i] ]][[ mynorms[j] ]]
    colnames(rawmat)=samp.labels; colnames(normmat)=samp.labels
    plotdata = list(plotdir='.',plotbase=paste(mynorms[j],mytypes[i],sep='.'),plottitle=proj.title)
    rowmask = rowSums(rawmat>1,na.rm=T) > (ncol(rawmat)/length(unique(grpBy)) ) & rowSums(is.na(normmat)) < (ncol(normmat)/length(unique(grpBy)) )
    if(sum(rowmask)>hclust.limit){ #hack! build optimization!
      rowmask = rowSums(rawmat>=quantile(rawmat,probs=hc.frac.cut,na.rm=T),na.rm=T) > (ncol(rawmat)/length(unique(grpBy)) ) & rowSums(is.na(normmat)) < (ncol(normmat)/length(unique(grpBy)) )
    }
    ans = summary.plots(rawmat=log2(rawmat +1), normmat=normmat, mynorm=mynorms[j], samp.labels=samp.labels, samp.classes=grpBy, plotdata=plotdata,plot2file=TRUE,histbins=histbins, colorspec=colors.rgb)
    ans = qc.clusters(rawmat=log2(rawmat[rowmask,] +1), normmat=normmat[rowmask,], attribs=annCol[annCol.plotme], oneclass=annCol.lmBy, colorspec=colors.rgb, plotdata=plotdata, plot2file=TRUE, clim.pct=clim.pct)
  }
}
```

### Optionally plot bias-reduced data filtered by prior data
```{r}

# loop over count types, bias reductions and prior data sets
# filter rows by prior data and regenerate heatmaps and MDS plots 
# confirm plotting constants from above
annCol.plotme = c(annCol.lmBy,"DV200","RNA Yield (ng)")
for(i in 1:length(mytypes) ){
  if( grepl('SJ',mytypes[i]) ){ #not the most robust way to find SJs
    # map SJ positions to overlapping genes
    # specifying defaults to fix "Gene" and Pos as colnames
    gtf.col = c("Gene","Symbol","Chr","start","stop") 
    pos.col = "Pos"
    mySJ_dt = mapSJ2feature(STAR.data$LoM.raw[[ mytypes[i] ]][[1]], pos.col=pos.col, gtf.col=gtf.col, gtf.file=gtfFile, gtf.Rdir="/home/exacloud/lustre1/BioCoders/ProjectCollaborations/PSchedin_RNA161024PS/code/GenomicsFunctions") # min data rows for heatmap and MDS plots
    # map gene IDs back to data matrix
    idx2 = match( rownames(STAR.data$LoM.raw[[ mytypes[i] ]][[1]]), mySJ_dt[,get(pos.col)] )
    idx1 = which(!is.na(idx2)); idx2 = idx2[idx1]
  }
  for(j in 1:length(mynorms)){
    # set up matrices and config for plotting
    normmat= LoM.norms[[ mytypes[i] ]][[ mynorms[j] ]]
    colnames(normmat)=samp.labels
    if( grepl('SJ',mytypes[i]) ){ #not the most robust way to find SJs
      # this depends on normmat and raw data having same row order
      rownames(normmat)[idx1] = mySJ_dt[idx2,get(gtf.col[1])]
    }
    # loop over prior data sets
    for(k in 1:length(priordata_lsdt) ){
    plotdata = list(plotdir='.',plotbase=paste(mynorms[j],mytypes[i],names(priordata_lsdt)[k],sep='.'),plottitle=paste(proj.title,names(priordata_lsdt)[k]) )
    rowmask = rownames(normmat) %chin% priordata_lsdt[[k]][,get(priordata.geneID)]
      ans = qc.clusters(rawmat=normmat[rowmask,], normmat=normmat[rowmask,], attribs=annCol[annCol.plotme], oneclass=annCol.lmBy, colorspec=colors.rgb, plotdata=plotdata, plot2file=TRUE, clim.pct=clim.pct)
    }
  }
}


```

### Run regression based on experimental design and/or technical parameters
```{r}
regress_lsls = vector(mode='list',length=length(STAR.data$LoM.raw))
names(regress_lsls) = names(STAR.data$LoM.raw)
contr_ls = list("ER Status"=list(baseline="positive",contr.FUN="contr.treatment"))
# set baseline for regression in parameter(s) of interest
# contr.treatment generates regression coefficients that are like (adjusted mean) ratios of other groups to the baseline group.
# contr.sum generates coefficients that are like (adjusted mean) ratios to average all for all but the mandadory ommitted treatment group (because there is always one fewer independent pairwise comparison than there are pairs).

lm_expr = "y ~ `ER Status`"
rowmask_ls = vector(mode='list',length=length(STAR.data$LoM.raw))
names(rowmask_ls) = names(STAR.data$LoM.raw)
for(i in 1:length(mytypes) ){
  for(j in which(!mynorms %in% do.not.regress) ){
    # set up data matrices
    if( length(STAR.data$LoM.raw[[ mytypes[i] ]])>1 ){
      rawmat = STAR.data$LoM.raw[[ mytypes[i] ]][[myreads]]
    } else {
      rawmat = STAR.data$LoM.raw[[ mytypes[i] ]][[1]]
    }
    tmp = unlist(lapply(contr_ls,function(x){x$baseline}))
    tmp=paste(names(tmp),tmp,sep=".")
    plotdata = list(plotdir='./',plotbase=paste(mynorms[j],mytypes[i],'vs',tmp,sep='.'),plottitle=proj.title)
    normmat= LoM.norms[[ mytypes[i] ]][[ mynorms[j] ]]
    colnames(rawmat)=samp.labels; colnames(normmat)=samp.labels
    # prepare rowmask for heatmap/MDS (remove non-expr or low expr>hclustlim)
    rowmask = rowSums(rawmat>1,na.rm=T) > (ncol(rawmat)/length(unique(grpBy)) ) & rowSums(is.na(normmat)) <= na.lim
    rowmask_ls[[ mytypes[i] ]][[ mynorms[j] ]] = rowmask # save for later
    # regression
    regress_lsls[[mytypes[i]]][[mynorms[j]]] = regressMatrix(normmat[rowmask,], expt.design=annCol[annCol.lmBy], lm_expression=lm_expr, contr_list = contr_ls, plot2file = TRUE, plotdata = plotdata)
  }
}
```

### preview results
```{r}
topn=500 # number with which to play
for( i in 1:length(mytypes) ){
  for( j in which(!mynorms %in% do.not.regress) ){
    ans = regress_lsls[[mytypes[i]]][[mynorms[j]]]$q_list
    cat(mytypes[i],mynorms[j],"1-p0:",signif(unlist(lapply(2:length(ans),function(x){1-ans[[x]]$pi0})),2),"\n")
    cat(mytypes[i],mynorms[j],"qcut:",signif(unlist(lapply(2:length(ans),function(x){quantile(ans[[x]]$qvalues,probs=topn/length(ans[[x]]$qvalues))})),3),"returns ~",topn,"out of",length(ans[[2]]$qvalues),"\n")
}}
```


# Plots for each design factor (default) or for factors specified in facSel
#   1) Histogram of p values that were included in the design
#      Look for a peak on the left, and no peaks in the middle or on the right
#   2) qvalue's default plots, with full qvalue range c(0,1) plotted
#      Look for descent to pi0 in the top left tuning plot with good asymptote
#      The slow/steep rise in q-values in remaining plots depends on resolving power of data

```{r}
lmBy = annCol.lmBy
histbins=20
for( i in 1:length(mytypes) ){
  for( j in which(!mynorms %in% do.not.regress) ){
    normmat= LoM.norms[[ mytypes[i] ]][[ mynorms[j] ]]
    colnames(normmat)=samp.labels
    reg_ls = regress_lsls[[mytypes[i]]][[mynorms[j]]]
    mymask= rowmask_ls[[mytypes[i]]][[mynorms[j]]]
    # pull out regression design and set up plot config
    tmp = unlist(lapply(contr_ls,function(x){x$baseline}))
    tmp=paste(names(tmp),tmp,sep=".")
    plotdata = list(plotdir='./',plotbase=paste(mynorms[j],mytypes[i],tmp,sep='.'),plottitle=proj.title)
    for(fac in names(reg_ls$q_list)[grep(lmBy,names(reg_ls$q_list))] ){
      pdata = plotdata
      pdata$plotbase = paste(plotdata$plotbase,make.names(fac),sep='.')
      ans = qcQvalues(norm_x=normmat[mymask,], pvalue_v=reg_ls$p_mat[,fac], obj_qvalue=reg_ls$q_list[[fac]], attribs=annCol[lmBy], oneclass=lmBy, plotdata=pdata, colorspec=colors.rgb, histbins=histbins, plot2file=TRUE)
    }
  }
}
```

# Build of ratios to baselines for desired design factors 
# Also build masks selecting genes based on q-values per factor 
#  and follow-up considerations such as expression level and fold change
# Loop over q-value cuts to assist with final cut selection.
# Plot heatmaps and MDS plots based on selected genes and designed ratios

```{r}
# cuts
ngene_v = c(200,500,1000) # q-value cuts by number; can also cut by q-value
ratioby_ls = list("ER Status"=contr_ls$`ER Status`$baseline)
ratio_fold = 1.3
intensity_fold = 2
priordata_IDs = unique(unlist(lapply(priordata_lsdt,function(x){x[,get(priordata.geneID)]})))
cut_ls = list(q_combine="OR", rcut_fold=ratio_fold, icut_fold=intensity_fold, include_ID=priordata_IDs)
# settings for heatmaps
annCol.plotme = c(annCol.lmBy,"DV200","RNA Yield (ng)") # heatmap tracks
clustrowmin = 10 # min data rows for heatmap and MDS plots
# save ratios and selections
select_lsmk = vector(mode='list',length=length(STAR.data$LoM.raw))
names(select_lsmk) = names(STAR.data$LoM.raw)
ratio_lsmat = vector(mode='list',length=length(STAR.data$LoM.raw))
names(ratio_lsmat) = names(STAR.data$LoM.raw)
for( i in 1:length(mytypes) ){
  for( j in which(!mynorms %in% do.not.regress) ){
    normmat= LoM.norms[[ mytypes[i] ]][[ mynorms[j] ]]
    colnames(normmat)=samp.labels
    if( grepl('SJ',mytypes[i]) ){ #not the most robust way to find SJs
      # map gene IDs back to data matrix using mapping built earlier
      # this enables include_ID to select rows
      idx2 = match( rownames(STAR.data$LoM.raw[[ mytypes[i] ]][[1]]), mySJ_dt[,get(pos.col)] )
      idx1 = which(!is.na(idx2)); idx2 = idx2[idx1]
      rownames(normmat)[idx1] = mySJ_dt[idx2,get(gtf.col[1])]
    }
    mymask= rowmask_ls[[mytypes[i]]][[mynorms[j]]]
    reg_ls = regress_lsls[[mytypes[i]]][[mynorms[j]]]$q_list
    reg_ls = reg_ls[!grepl('Intercept',names(reg_ls))]
    plotdata = list(plotdir='./',plotbase=paste(mynorms[j],sub('counts','ratios',mytypes[i]),'minratio',paste0(ratio_fold,'x'),'minexpr',round(min(normmat,na.rm=T)+log2(intensity_fold),1),sep='.'),plottitle=proj.title)
    for( ngenes in ngene_v ) {
      # calculate and plot ratios
      cut_ls$qcut = ngenes
      pdata = plotdata
      pdata$plotbase = paste(plotdata$plotbase,'ratio.vs',ratioby_ls$strain,ngenes,sep='.')
      # this function makes the ratios and cuts
      ans = designRatios(normmat[mymask,], q_list=reg_ls, attribs=annCol[annCol.plotme], ratioby_ls=ratioby_ls, cut_ls=cut_ls)
      # this function creates the heatmap and MDS plot
      if( sum(ans$rowmask)>clustrowmin ){
        ans2= plotRatios( ratiomat=ans$ratiomat, attribs=annCol[annCol.plotme], oneclass=lmBy, plotdata=pdata, colorspec=colors.rgb, rowmask=ans$rowmask, plot2file=TRUE)
      }
      # save selections
      select_lsmk[[mytypes[i]]][[mynorms[j]]][paste(fac,ngenes,sep=":")] = ans['rowmask']
    }
    # save ratiomat (not dependent on ngene cut) after ngene loop
    ratio_lsmat[[mytypes[i]]][[mynorms[j]]] = ans$ratiomat
  }
}
```


###Produce Scatter plots based on lmBy
```{r}
for( i in 1:length(mytypes) ){
  for( j in 1:length(mynorms) ){
    scatter_mat= LoM.norms[[ mytypes[i] ]][[ mynorms[j] ]]
    colnames(normmat)=samp.labels
    plotdata = list(plotdir='./',plotbase=paste(mynorms[j],mytypes[i],'Scatter',lmBy,sep=':'),plottitle=proj.title)
    scatterplot(scatter_mat, attribs=annCol[[lmBy]],plotdata=plotdata,plot2file = TRUE) 
  }
}
```


#start with gene counts, loess norm
#translate to human for annotation purposes
```{r}
i=1; j=which(mynorms=="loess")
ratio_mat = ratio_lsmat[[mytypes[i]]][[mynorms[j]]]
reg_ls = regress_lsls[[mytypes[i]]][[mynorms[j]]]
ID = rownames(ratio_mat)
```


###Human numeric EntrezGene IDs
#Change the srcSpecies to the appropriate species
```{r}
hsa_entrezID = inpIDMapper(ID, srcSpecies="RATNO", destSpecies="HOMSA", srcIDType="ENSEMBL", destIDType="EG", keepMultDestIDMatches = FALSE)
Entrez.ID = character(length(ID))
Entrez.ID[] = NA
idx2 = match(ID,names(hsa_entrezID))
idx1 = which(!is.na(idx2)); idx2= idx2[idx1]
Entrez.ID[idx1] = hsa_entrezID[idx2];
Entrez.ID = as.numeric(Entrez.ID)
```


###Human gene symbols
```{r}
hsa_symbol = inpIDMapper(ID, srcSpecies="RATNO", destSpecies="HOMSA", srcIDType="ENSEMBL", destIDType="SYMBOL", keepMultDestIDMatches = FALSE)
Anno.Symbol = character(length(ID))
Anno.Symbol[] = NA
idx2 = match(ID,names(hsa_symbol))
idx1 = which(!is.na(idx2)); idx2= idx2[idx1]
Anno.Symbol[idx1] = unlist(hsa_symbol[idx2]);
# assemble IDs for annotation
backgroundset = as.data.table(cbind(ID, Entrez.ID, Anno.Symbol))
# assemble signatures
```


```{r}
ngenes = 3000
fac = names(reg_ls$q_list)[grep(lmBy,names(reg_ls$q_list))]
rowmask = select_lsmk[[mytypes[i]]][[mynorms[j]]][[paste(fac,ngenes,sep=":")]]
sig_gmt = NULL
betas = reg_ls$b_mat[,fac]
sig_gmt[['all']] =  ID[rowmask]
sig_gmt[['up']] = ID[rowmask & betas>0]
sig_gmt[['dn']] = ID[rowmask & betas<=0]
```


###Annotate gene sets with default anno sets
```{r}
fileSettings = list(directory='/PATH/TO/PROJECT/DIRECTORY/',baseFilename="Pathway_top3K_Trial_Pressure")
path_ls = ags.wrapper(setlist=sig_gmt, backgroundset, include.identifiers=TRUE, anno.uni=TRUE, 
                      fileSettings = fileSettings,
                      functiondir = "/PATH/TO/AssociationFunctions/",
                      resourcedir = "PATH/TO/AssocationData",
                      return.OM=TRUE, ecut=0.05, ocut=5)

fileSettings = list(directory='/PATH/TO/PROJECT/DIRECTORY/',baseFilename="Pathway_top3K_Trial_Pressure_noUNI")

```


###Annotate gene sets with default anno sets BUT restrict uni to overlap
```{r}
pathi_ls = ags.wrapper(setlist=sig_gmt, backgroundset, include.identifiers=TRUE, anno.uni=FALSE,    # <
                       fileSettings = fileSettings,
                       functiondir = "/PATH/TO/AssociationFunctions/",
                       resourcedir = "PATH/TO/AssocationData",
                       return.OM=TRUE, ecut=0.05, ocut=5)


out_norm_mat = LoM.norms$gene.counts$loess[rowmask_ls$gene.counts$loess,]
out_table = outputTable(normmat= out_norm_mat, gtf.file = gtfFile,ratiomat = ratio_lsmat$gene.counts$loess, q_list=reg_ls$q_list)
write.csv(out_table, row.names = FALSE, file='LoessNormed_with_Ratio_trialdata.txt',quote=FALSE)
```
